### PR TITLE
Dominator

### PR DESCRIPTION
**Pull request by serpilliere**
Dominator algorithm can only work on graph with one head.
Add PostDominator algorithm

### CODE REVIEW

File: miasm2/core/graph.py [hunk 1]
```diff
@@ -162,42 +148,113 @@ def dot(self):
...
+                continue
+            reachable.add(node)
+            yield node
+            for pred in self.predecessors_iter(node):
+                todo.add(pred)
+
+    def compute_dominators(self, head):
+        """Compute dominators of the graph"""
+        nodes = set(self.reachable_sons(head))
         dominators = defaultdict(set)
```
Comment (commial) [inline miasm2/core/graph.py]: Where did you use the fact that dominators is a `defaultdict`?
You can use `set` for copy in line 184, and avoid using again a `dict()` at the end of the function. In addition, `defaultdict` can easily be prone to error.

File: miasm2/core/graph.py [hunk 2]
```diff
@@ -162,42 +148,113 @@ def dot(self):
         return out
 
 
+    def reachable_sons(self, head):
```
Comment (commial) [inline miasm2/core/graph.py]: To factorize and avoid future copy/paste error on patches, you could use a protected method with a callback (`successors_iter` for `reachable_sons` and `predecessors_iter` for `reachable_parents`)

File: miasm2/core/graph.py [hunk 3]
```diff
@@ -162,42 +148,113 @@ def dot(self):
...
                 continue
+
             dominators[node] = new_dom
-            for succ in self.successors(node):
+            for succ in self.successors_iter(node):
                 todo.add(succ)
         return dict(dominators)
+
+
+    def compute_postdominators(self, leaf):
```
Comment (commial) [inline miasm2/core/graph.py]: Same remark as `reacheable_*`. As dominator and post-dominator algorithms are completely symmetric, you can use two callbacks for the only changes: `reachable_*` and `*_iter`.

**Comment by serpilliere** [comment]
Correct.
The default dict is removed.
The code is factorized for dominators/postdominators.

**Comment by commial** [comment]
Thanks for your modifications. Did you intend to apply code factorizing on `reacheable_*` too ?

**Comment by serpilliere** [comment]
Correct.
Code factorized.

File: miasm2/core/graph.py [hunk 4]
```diff
@@ -162,42 +148,89 @@ def dot(self):
...
+            if node in reachable:
+                continue
+            reachable.add(node)
+            yield node
+            for next_node in next_cb(node):
+                todo.add(next_node)
+
+    def reachable_sons(self, head):
+        """Compute every nodes reachable from node @head"""
+        for node in self._reachable_nodes(head, self.successors_iter):
```
Comment (commial) [inline miasm2/core/graph.py]: `self._reachable_nodes(head, self.successors_iter)` is already an iterator. You can directly write:
`return self._reachable_nodes(head, self.successors_iter)`.

For instance:
```Python
def test():
    for i in xrange(4):
        yield i
def foobar():
    return test()
for i in foobar():
    print i
# 0
# 1
# 2
# 3
```

Idem for `reachable_parents` :smiley:

**Comment by serpilliere** [comment]
Modified.

**Pull request merged**